File: app/src/main/java/com/example/shopease/LoginActivity.java
package com.example.shopease;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

public class LoginActivity extends AppCompatActivity {
    private FirebaseAuth auth;
    private FirebaseDatabase database;
    private boolean isLoggingIn = false; // Prevents multiple login requests
    private ProgressBar progressBar; // Loading indicator

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.login_screen);

        auth = FirebaseAuth.getInstance();
        database = FirebaseDatabase.getInstance();

        // UI Elements
        EditText emailField = findViewById(R.id.emailField);
        EditText passwordField = findViewById(R.id.passwordField);
        Button loginButton = findViewById(R.id.loginButton);
        TextView signUpText = findViewById(R.id.signUpText);
        progressBar = findViewById(R.id.progressBar); // Add a ProgressBar in your layout

        // Handle Login button click
        loginButton.setOnClickListener(v -> {
            String email = emailField.getText().toString().trim();
            String password = passwordField.getText().toString().trim();

            if (!email.isEmpty() && !password.isEmpty()) {
                loginUser(email, password);
            } else {
                Toast.makeText(this, "Please fill in all fields", Toast.LENGTH_SHORT).show();
            }
        });

        // Handle SignUp navigation
        signUpText.setOnClickListener(v -> startActivity(new Intent(this, RegistrationActivity.class)));
    }

    private void loginUser(String email, String password) {
        if (isLoggingIn) return; // Prevent multiple logins
        isLoggingIn = true;
        progressBar.setVisibility(View.VISIBLE); // Show loading indicator

        auth.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(task -> {
                    isLoggingIn = false;
                    progressBar.setVisibility(View.GONE); // Hide loading indicator

                    if (task.isSuccessful()) {
                        FirebaseUser user = auth.getCurrentUser();
                        if (user != null) {
                            navigateBasedOnUserRole(user.getUid());
                        } else {
                            Toast.makeText(this, "User not found", Toast.LENGTH_SHORT).show();
                        }
                    } else {
                        Toast.makeText(this, "Login failed: " + task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void navigateBasedOnUserRole(String userId) {
        DatabaseReference userRef = database.getReference("Users").child(userId).child("role");

        userRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    String role = snapshot.getValue(String.class);
                    if (role != null) {
                        Intent intent;
                        switch (role) {
                            case "user":
                                intent = new Intent(LoginActivity.this, HomeActivity.class);
                                break;
                            case "seller":
                                intent = new Intent(LoginActivity.this, SellerDashboardActivity.class);
                                break;
                            case "admin":
                                intent = new Intent(LoginActivity.this, AdminDashboardActivity.class);
                                break;
                            default:
                                Toast.makeText(LoginActivity.this, "Unknown role", Toast.LENGTH_SHORT).show();
                                return;
                        }
                        startActivity(intent);
                        finish();
                    } else {
                        Toast.makeText(LoginActivity.this, "User role not found", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(LoginActivity.this, "User data missing", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Toast.makeText(LoginActivity.this, "Failed to retrieve user role: " + error.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/shopease/LoginActivity.kt
package com.example.shopease

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.example.shopease.seller.SellerDashboardActivity
import com.example.shopease.admin.AdminDashboardActivity

class LoginActivity : AppCompatActivity() {
    private lateinit var auth: FirebaseAuth
    private lateinit var database: FirebaseDatabase

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.login_screen)

        // Initialize FirebaseAuth and FirebaseDatabase
        auth = FirebaseAuth.getInstance()
        database = FirebaseDatabase.getInstance()

        val emailField = findViewById<EditText>(R.id.emailField)
        val passwordField = findViewById<EditText>(R.id.passwordField)
        val loginButton = findViewById<Button>(R.id.loginButton)
        val signUpText = findViewById<TextView>(R.id.signUpText)

        // Handle login button click
        loginButton.setOnClickListener {
            val email = emailField.text.toString().trim()
            val password = passwordField.text.toString().trim()

            // Validate input fields
            if (email.isNotEmpty() && password.isNotEmpty()) {
                loginUser(email, password)
            } else {
                Toast.makeText(this, "Please fill in all fields", Toast.LENGTH_SHORT).show()
            }
        }

        // Handle SignUp Text click to navigate to RegistrationActivity
        signUpText.setOnClickListener {
            startActivity(Intent(this, RegistrationActivity::class.java)) // Navigate to RegistrationActivity
        }
    }

    private fun loginUser(email: String, password: String) {
        // Sign in with Firebase Authentication
        auth.signInWithEmailAndPassword(email, password)
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    val userId = auth.currentUser?.uid
                    navigateBasedOnUserRole(userId)
                } else {
                    Toast.makeText(this, "Login failed: ${task.exception?.message}", Toast.LENGTH_SHORT).show()
                }
            }
    }

    private fun navigateBasedOnUserRole(userId: String?) {
        if (userId == null) return

        database.getReference("Users").child(userId).child("role").get()
            .addOnSuccessListener { snapshot ->
                when (snapshot.value.toString()) {
                    "user" -> startActivity(Intent(this, HomeActivity::class.java)) // Navigate to HomeActivity
                    "seller" -> startActivity(Intent(this, SellerDashboardActivity::class.java)) // Navigate to Seller Dashboard
                    "admin" -> startActivity(Intent(this, AdminDashboardActivity::class.java)) // Navigate to Admin Dashboard
                    else -> Toast.makeText(this, "Unknown role", Toast.LENGTH_SHORT).show()
                }
                finish() // Close LoginActivity after navigating
            }
    }
}
